### PR TITLE
Make Caitlin's BGEN PRS Pipeline Fast

### PR DESCRIPTION
**Pull request by danking**
This needs tests as well as I think I need to fix indexing (so I don't blow memory on the full genome), but I wanted to share what I've been up to with y'all. Also, caitlin can use this branch to run an analysis if it comes to that. I would also appreciate some feedback on the approach. It would be much more ideal to just make joins of variant tables against BGENs smarter, but I think the infrastructure necessary for that is big.

cc: @cseed

List of changes:

 - added `_variants_per_file` limits the loaded variants to variants at the given array of indexes (0-indexed, same order as on disk)

 - ~added `row_fields` which prevents reading and allocation of LID and RSID (also improved python-type-checking for `row_fields` and `entry_fields`)~ Moved to #3779 and #3778

 - ~fixed table-table joins to _not_ always coerce (thus computing partition keys of) the right-hand table~ Moved to #3723 

 - ~added a check that prevents globals and sample annotations copying when they're not used in the body of a MatrixMapCols~ Moved to #3751

 - ~fixed a bug in `IndexBTree` wherein if the number of elements was a multiple of 1024, an unnecessary 1024 elements were added to the end of the index file (which I believe breaks the reading process which expects the number of bytes to correspond to the size of the tree)~ Moved to #3750

 - ~added `IndexBTree2` which is just an in-memory list of the variant start positions. This is a fair bit of data. Chromosome 1 has about 250 million bases, so in the worst case this is 250 * 8 million bytes = 2 GB. It occurs to me that this is actually way to much data to load on the master node in general (since I just try to open the indexes for every file). I should switch this to a disk-based index.~ Made it disk-based, called it `OnDiskBTreeIndexToValue`  #3794

 - each hadoop `FileSplit` now contains a possibly null (indicating no filter) list of variants (by index) to keep, in practice this should be quite small

 - ~I changed several asserts to `if`'s with fatals, so as not to allocate strings~ Moved to #3771

 - ~We no longer copy the genotype data into a buffer in the block reader. This was forcing the `fastKeys` to do an unnecessary data copy~ Moved to #3783 (with some substantial refactoring so it doesn't look much like this PR anymore)

 - ~I changed the contract on BgenRecord to require that `getValue` is called to "consume" the record before the next record is taken~ Irrelevant thanks to #3783 's refactoring

 - ~`getValue(null)` just skips bytes (no copy, no decompression)~ Irrelevant thanks to #3783 's refactoring

 - ~I added `RegionValueBuilder.unsafeAdvance` which can be used when you're creating an array of empty structs but don't want to do all the unnecessary RVB bookkeeping work.~ Moved to #3773

 - ~I use `RegionValueBuilder.unsafeAdvance` to make loading a BGEN without entry fields very fast.~ Rolled into #3783

 - ~I fixed `Table.index` to not trigger a partition key info gathering~ Moved to #3774

I had to ship the arrays of filtered variant indices to the workers somehow, so I shipped them as base64 encoded arrays of bytes. It's pretty groady (and that's why I added the commons-codec library). I don't know how else to initialize record readers with hadoop.

Generally, I think the BGEN loading code could use a clean up, and I haven't done that here, if anything I've made it more complicated.

I also need to check that there are tests for or write tests for:

 - indexing tables doesn't cause an extra shuffle
 - ~the include lid and include raid flags~ included in #3779
 - the variant list flag
 - `getSplits`
 - the variant filtering code
 - ~loading a bgen with no entries~ exists: `BGENTests.test_import_bgen_no_entries`

### CODE REVIEW

File: python/hail/table.py [hunk 1]
```diff
@@ -1291,7 +1291,7 @@ def types_compatible(left, right):
                 analyze('Table.index', e, src._row_indices)
 
             is_key = src.key is not None and len(exprs) == len(src.key) and all(
-                expr is key_field for expr, key_field in zip(exprs, self.key.values()))
+                expr is key_field for expr, key_field in zip(exprs, src.key.values()))
```
Comment (jigold) [inline python/hail/table.py]: Didn't we already merge this fix? Why is it showing up in the diff?

File: src/main/scala/is/hail/expr/Relational.scala [hunk 2]
```diff
@@ -1402,7 +1402,13 @@ case class MatrixMapCols(child: MatrixIR, newCol: IR, newKey: Option[IndexedSeq[
     val colValuesType = TArray(prev.typ.colType)
     val vaType = prev.typ.rvRowType
 
+    var initOpNeedsSA = false
+    var initOpNeedsGlobals = false
+
     val transformInitOp: (Int, IR) => IR = { (nAggs, initOpIR) =>
+      initOpNeedsSA = Mentions(initOpIR, "sa")
```
Comment (jigold) [inline src/main/scala/is/hail/expr/Relational.scala]: I think you should break this out into a separate PR. It's a good change that would be nice to have irrespective of BGEN.

File: src/main/scala/is/hail/io/IndexBTree.scala [hunk 3]
```diff
@@ -32,7 +32,8 @@ object IndexBTree {
       layers.append(arr)
 
       // Pad last layer so last block is 1024 elements (1024*8 bytes)
-      val paddingRequired = 1024 - (arr.length % 1024)
+      val danglingElements = (arr.length % 1024)
+      val paddingRequired = if (danglingElements == 0) 0 else 1024 - danglingElements
```
Comment (jigold) [inline src/main/scala/is/hail/io/IndexBTree.scala]: Can you make this a separate PR?

**Comment by cseed** [comment]
Instead of `_include_*`, you should have `row_fields=['lid', 'rsid']`.  `import_bgen` should produce a matrix read IR whose type is determined by the entry and row fields, and the Hadoop flags should get set based on the fields in the type so it can play nicely with, e.g., @tpoterba's coming prune fields PR.

File: python/hail/table.py (hunk 1 above)
Comment (danking) [reply python/hail/table.py]: not sure. it's definitely fixed in master.

**Comment by danking** [comment]
@cseed, done.

@jigold done & done. Also rebasing removed the `table.py` change

**Comment by jigold** [comment]
@danking Sorry to keep making you break things out, but it is really helpful for me and the changes will go in faster. Can you make a separate PR with the following changes that don't relate to passing the indices and the new index code? Specifically, the following items from your list:

```
added row_fields which prevents reading and allocation of LID and RSID (also improved python-type-checking for row_fields and entry_fields)

I changed several asserts to if's with fatals, so as not to allocate strings

We no longer copy the genotype data into a buffer in the block reader. This was forcing the fastKeys to do an unnecessary data copy

I changed the contract on BgenRecord to require that getValue is called to "consume" the record before the next record is taken

getValue(null) just skips bytes (no copy, no decompression)

I added RegionValueBuilder.unsafeAdvance which can be used when you're creating an array of empty structs but don't want to do all the unnecessary RVB bookkeeping work.

I use RegionValueBuilder.unsafeAdvance to make loading a BGEN without entry fields very fast.

I fixed Table.index to not trigger a partition key info gathering
```

File: src/main/scala/is/hail/io/bgen/BgenRecord.scala [hunk 4]
```diff
@@ -31,109 +68,137 @@ class BgenRecordV12(compressed: Boolean, nSamples: Int,
...
+        val c2 = Call2.fromUnphasedDiploidGtIndex(2)
+
+        var i = 0
+        while (i < nSamples) {
+          val off = nSamples + 10 + 2 * i
+          val d0 = bytes(off) & 0xff
+          val d1 = bytes(off + 1) & 0xff
+
+          if ((bytes(8 + i) & 0x3f) != 2)
+            fatal(s"Ploidy value must equal to 2. Found ${(bytes(8 + i) & 0x3f)}")
```
Comment (jigold) [inline src/main/scala/is/hail/io/bgen/BgenRecord.scala]: format: `${ (bytes(8 + i) & 0x3f) }`

File: src/main/scala/is/hail/io/bgen/BgenRecord.scala [hunk 5]
```diff
@@ -31,109 +68,137 @@ class BgenRecordV12(compressed: Boolean, nSamples: Int,
...
+
+          if ((bytes(8 + i) & 0x80) != 0)
```
Comment (jigold) [inline src/main/scala/is/hail/io/bgen/BgenRecord.scala]: Why are you reading the bytes above if the sample is missing? I think it was better in the original order.

File: src/main/scala/is/hail/io/bgen/BgenRecord.scala [hunk 6]
```diff
@@ -1,9 +1,12 @@
 package is.hail.io.bgen
 
+import com.google.common.io.LittleEndianDataInputStream
```
Comment (jigold) [inline src/main/scala/is/hail/io/bgen/BgenRecord.scala]: I can't tell where this is used. Also, can you explain what the new Inflater, InflaterInputStream classes are doing?

File: src/main/scala/is/hail/io/IndexBTree.scala [hunk 7]
```diff
@@ -138,3 +138,40 @@ class IndexBTree(indexFileName: String, hConf: Configuration, branchingFactor: I
...
+      1
+    else
+      (math.log(size) / math.log(1024)).ceil.toInt
+  }
+
+  private[this] def leadingElements(layer: Int): Long = {
+    var i = 0
+    var leadingElements = 0L
+    while (i < layer - 1) {
+      leadingElements = leadingElements * 1024L + 1024L
```
Comment (jigold) [inline src/main/scala/is/hail/io/IndexBTree.scala]: How did you derive this? I'm having trouble determining if it's correct.

File: src/main/scala/is/hail/HailContext.scala [hunk 8]
```diff
@@ -299,23 +299,29 @@ class HailContext private(val sc: SparkContext,
...
     includeDosage: Boolean,
+    includeLid: Boolean,
+    includeRsid: Boolean,
     nPartitions: Option[Int] = None,
     rg: Option[ReferenceGenome] = Some(ReferenceGenome.defaultReference),
     contigRecoding: Option[Map[String, String]] = None,
     skipInvalidLoci: Boolean = false): MatrixTable = {
-    importBgens(List(file), sampleFile, includeGT, includeGP, includeDosage, nPartitions, rg,
-      contigRecoding, skipInvalidLoci)
+    importBgens(List(file), sampleFile, includeGT, includeGP, includeDosage, includeLid, includeRsid,
```
Comment (jigold) [inline src/main/scala/is/hail/HailContext.scala]: Can you add the includedVariantsPerFile here as well.

**Comment by danking** [comment]
Replaced by all the mentioned PRs and https://github.com/hail-is/hail/pull/3813
